app/models/status.rb

```diff
@@ -188,10 +191,21 @@ def as_home_timeline(account)
       where(account: [account] + account.following).where(visibility: [:public, :unlisted, :private])
     end
 
-    def as_direct_timeline(account)
-      query = joins("LEFT OUTER JOIN mentions ON statuses.id = mentions.status_id AND mentions.account_id = #{account.id}")
-              .where("mentions.account_id = #{account.id} OR statuses.account_id = #{account.id}")
-              .where(visibility: [:direct])
+    def as_direct_timeline(account, max_id = nil, since_id = nil, limit = 20)
+      dm_to_me = Status.direct_messages.message_to_me(account).select(:id).limit(limit)
+      dm_from_me = Status.direct_messages.message_from_me(account).select(:id).limit(limit)
+
+      if max_id.present?
+        dm_to_me = dm_to_me.where(arel_table[:id].lt(max_id))
+        dm_from_me = dm_from_me.where(arel_table[:id].lt(max_id))
+      end
+
+      if since_id.present?
+        dm_to_me = dm_to_me.where(arel_table[:id].gt(since_id))
+        dm_from_me = dm_from_me.where(arel_table[:id].gt(since_id))
+      end
+
+      query = Status.where(id: (dm_to_me + dm_from_me).uniq)
 
       apply_timeline_filters(query, account, false)
```

Comment:
This query does  limit → filter → limit.
If filter removes rows, the result row count may be less than expected limit.

example:
_from_me returns 20 rows (clipped by limit)
_to_me returns 4 rows (there is no more row)
merged row is 22 (2 rows removed by uniq)
filter removes 6 rows
result is 16 rows. (there is more rows, but result is less than expected)

more extreme case, you may get 0 rows even if  there is more rows.